src/azure-cli-core/azure/cli/core/_profile.py

```diff
@@ -861,5 +861,8 @@ def _create_identity_instance(cli_ctx, *args, **kwargs):
     from .telemetry import set_broker_info
     set_broker_info(allow_broker=allow_broker)
 
+    # PREVIEW: In Azure Stack environment, use core.instance_discovery=false to disable MSAL's instance discovery.
+    instance_discovery = cli_ctx.config.getboolean('core', 'instance_discovery', True)
```

Comment:
```suggestion
    instance_discovery = cli_ctx.config.getboolean('core', 'instance_discovery', None)
```

I would recommend let the default value match [MSAL's default value](https://msal-python.readthedocs.io/en/latest/#msal.ClientApplication.params.instance_discovery), rather than hardcoding a `True`. The benefit is, when/if MSAL changes the default behavior to `False` in the future, Azure CLI can automatically pick up the new behavior.

Reply:
`getboolean` doesn't support `None` as fallback. The fallback value must be a `bool`: https://github.com/microsoft/knack/blob/e496c9590792572e680cb3ec959db175d9ba85dd/knack/config.py#L146-L147

Reply:
Well, 5 years ago, I used to also design boolean flags as boolean type (such as [MSAL's old `validate_authority`](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/msal/application.py#L182)). Nowadays, [all new boolean-like parameters accept a `None` as default value](https://github.com/AzureAD/microsoft-authentication-library-for-python/blame/dev/msal/application.py#L194-L196). In other words, the *"default: let the upstream library do whatever it chooses to do"* becomes a legit input. The benefit of this new approach is that it no longer forces the downstream libraries/apps to pick a side, when they shouldn't.

Perhaps you can consider building a new `getboolean2(...)` which allows `None` as default, and potentially transition to replace the old `getboolean()` with the new one. `None` is equivalent to `False` when a real boolean is needed, anyway.

Reply:
This requires changing the underlying framework Knack which is a little bit heavy. Let's keep it the way it is (for now).